pkg/server/internal/packetio.go

```diff
@@ -155,13 +155,15 @@ func (p *PacketIO) readOnePacket() ([]byte, error) {
 		}
 		p.compressedSequence++
 		p.compressedWriter.compressedSequence = p.compressedSequence
+		compressedLength := int(uint32(compressedHeader[0]) | uint32(compressedHeader[1])<<8 | uint32(compressedHeader[2])<<16)
 		uncompressedLength := int(uint32(compressedHeader[4]) | uint32(compressedHeader[5])<<8 | uint32(compressedHeader[6])<<16)
 
 		if uncompressedLength > 0 {
 			switch p.compressionAlgorithm {
 			case mysql.CompressionZlib:
 				var err error
-				r, err = zlib.NewReader(p.bufReadConn)
+				lr := io.LimitReader(p.bufReadConn, int64(compressedLength))
```

Comment:
What about zstd?

Reply:
Unfortunately we can't use the same test for ZSTD as MySQL Connector/J doesn't support this: https://bugs.mysql.com/bug.php?id=112732

I created a similar test for this in C:
```c
#include <mysql.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#define COMPRESS_NONE 0
#define COMPRESS_ZLIB 1
#define COMPRESS_ZSTD 2

void test_compression(int compression, size_t payload_size) {
  static MYSQL *conn;
  MYSQL_STMT *stmt;
  MYSQL_BIND bind[1];
  MYSQL_RES *res;
  MYSQL_ROW row;
  long unsigned int length;

  conn = mysql_init(NULL);
  if (conn == NULL) {
    fprintf(stderr, "mysql_init() failed\n");
    exit(1);
  }

  switch (compression) {
  case COMPRESS_NONE:
    printf("testing with no compression and a payload of %zu.\n", payload_size);
    break;
  case COMPRESS_ZLIB:
    printf("testing with ZLIB and a payload of %zu.\n", payload_size);
    mysql_options(conn, MYSQL_OPT_COMPRESSION_ALGORITHMS, "zlib");
    break;
  case COMPRESS_ZSTD:
    printf("testing with ZSTD and a payload of %zu.\n", payload_size);
    mysql_options(conn, MYSQL_OPT_COMPRESSION_ALGORITHMS, "zstd");
    break;
  default:
    fprintf(stderr, "Error: no valid compression method specified.\n");
    exit(1);
  }
  uint ssl_mode = SSL_MODE_DISABLED;
  mysql_options(conn, MYSQL_OPT_SSL_MODE, &ssl_mode);

  if (mysql_real_connect(conn, "127.0.0.1", "root", "", "test", 4000, NULL,
                         0) == NULL) {
    fprintf(stderr, "Failed to connect to the database.\n");
    if (conn != NULL)
      fprintf(stderr, "Error: %s\n", mysql_error(conn));
    mysql_close(conn);
    exit(1);
  }

  const char *sql_to_send = "INSERT INTO BLOBTEST(blobdata) VALUES (?)";
  size_t required_size =
      payload_size - strlen(sql_to_send) - strlen("_binary ''");

  char buf[required_size];
  for (size_t i = 0; i < required_size; i++) {
    buf[i] = 65 + rand() % 26;
  }
  memset(bind, 0, sizeof(bind));
  bind[0].buffer_type = MYSQL_TYPE_BLOB;
  bind[0].length = &length;
  bind[0].is_null = 0;

  mysql_query(conn, "SET GLOBAL max_allowed_packet=6291456"); //  1024 * 1024 * 6
  mysql_query(conn, "DROP TABLE IF EXISTS BLOBTEST");
  mysql_query(conn, "CREATE TABLE BLOBTEST (pos int PRIMARY KEY "
                    "auto_increment, blobdata LONGBLOB)");

  stmt = mysql_stmt_init(conn);
  if (mysql_stmt_prepare(stmt, sql_to_send, strlen(sql_to_send))) {
    printf("Failed to prepare statement: %s\n", mysql_stmt_error(stmt));
    return;
  }
  if (mysql_stmt_bind_param(stmt, bind)) {
    printf("Failed to bind argument for prepared statement: %s\n",
           mysql_stmt_error(stmt));
    return;
  };
  size_t pos = 0;
  while (pos < strlen(buf)) {
    char sbuf[4096];
    size_t slen = sizeof(sbuf);
    if ((pos + slen) > strlen(buf)) {
	    slen = strlen(buf) - pos;
    }
    // printf("pos=%zu, strlen(buf)=%zu, slen=%zu\n", pos, strlen(buf), slen);
    memcpy(sbuf, buf + pos, slen);
    if (mysql_stmt_send_long_data(stmt, 0, sbuf, sizeof(sbuf))) {
      printf("Failed to send long data for prepared statement: %s\n",
             mysql_stmt_error(stmt));
      return;
    };
    pos += sizeof(sbuf);
  }

  if (mysql_stmt_execute(stmt)) {
    printf("Failed to execute prepared statement: %s\n",
           mysql_stmt_error(stmt));
    return;
  };

  if (mysql_query(conn, "SELECT blobdata from BLOBTEST LIMIT 1")) {
    printf("Failed to query blobdata: %s\n", mysql_error(conn));
    return;
  }

  res = mysql_store_result(conn);
  row = mysql_fetch_row(res);
  if (row != NULL) {
    if (strncmp(row[0], buf, strlen(buf)) != 0) {
      fprintf(stderr, "Buffer and result don't match.\n");
      return;
    }
  }

  mysql_close(conn);
}

int main() {
  srand(time(NULL));
  printf("Testing compression\n");

  test_compression(COMPRESS_NONE, 1024 * 1024 * 3 - 2);
  test_compression(COMPRESS_NONE, 1024 * 1024 * 3 - 1);
  test_compression(COMPRESS_NONE, 1024 * 1024 * 6);

  test_compression(COMPRESS_ZLIB, 1024 * 1024 * 3 - 2 - 3);
  test_compression(COMPRESS_ZLIB, 1024 * 1024 * 3 - 2 - 2);
  test_compression(COMPRESS_ZLIB, 1024 * 1024 * 3 - 1);
  test_compression(COMPRESS_ZLIB, 1024 * 1024 * 6);

  test_compression(COMPRESS_ZSTD, 1024 * 1024 * 3 - 2 - 3);
  test_compression(COMPRESS_ZSTD, 1024 * 1024 * 3 - 2 - 2);
  test_compression(COMPRESS_ZSTD, 1024 * 1024 * 3 - 1);
  test_compression(COMPRESS_ZSTD, 1024 * 1024 * 6);

  mysql_library_end();
  exit(0);
}
```

And running this against master (so without this PR) returns this:
```
$ gcc $(mysql_config --cflags --libs) -Wall -Wextra mysql-compression-test.c -std=gnu18 -g -o mysql-compression-test && ./mysql-compression-test 
Testing compression
testing with no compression and a payload of 3145726.
testing with no compression and a payload of 3145727.
testing with no compression and a payload of 6291456.
testing with ZLIB and a payload of 3145723.
Failed to send long data for prepared statement: Lost connection to MySQL server during query
testing with ZLIB and a payload of 3145724.
Failed to send long data for prepared statement: Lost connection to MySQL server during query
testing with ZLIB and a payload of 3145727.
Failed to send long data for prepared statement: Lost connection to MySQL server during query
testing with ZLIB and a payload of 6291456.
Failed to send long data for prepared statement: Lost connection to MySQL server during query
testing with ZSTD and a payload of 3145723.
testing with ZSTD and a payload of 3145724.
testing with ZSTD and a payload of 3145727.
testing with ZSTD and a payload of 6291456.
```

This is with this config:
```toml
[performance]
txn-entry-size-limit=125829120
```
I don't think the Java test required this, so the test might not be identical.

With this PR the output is like this:
```
$ gcc $(mysql_config --cflags --libs) -Wall -Wextra mysql-compression-test.c -std=gnu18 -g -o mysql-compression-test && ./mysql-compression-test 
Testing compression
testing with no compression and a payload of 3145726.
testing with no compression and a payload of 3145727.
testing with no compression and a payload of 6291456.
testing with ZLIB and a payload of 3145723.
testing with ZLIB and a payload of 3145724.
testing with ZLIB and a payload of 3145727.
testing with ZLIB and a payload of 6291456.
testing with ZSTD and a payload of 3145723.
testing with ZSTD and a payload of 3145724.
testing with ZSTD and a payload of 3145727.
testing with ZSTD and a payload of 6291456.
```

So looks like this doesn't affect ZSTD. 

Do you think I should still apply the same change for ZSTD, @djshow832 ?

Reply:
My intention is to send a packet longer than 16MB.